File: android/domain/src/main/kotlin/com/rideguard/domain/model/Screen.kt
package com.rideguard.domain.model

import kotlinx.serialization.Serializable

/**
 * Plain rectangle in screen pixels. Deliberately NOT android.graphics.Rect —
 * this module must stay Android-free so it can be tested on the JVM.
 */
@Serializable
data class Bounds(
    val left: Int,
    val top: Int,
    val right: Int,
    val bottom: Int,
) {
    val width: Int get() = right - left
    val height: Int get() = bottom - top
    val centerX: Int get() = (left + right) / 2
    val centerY: Int get() = (top + bottom) / 2
    val area: Int get() = maxOf(0, width) * maxOf(0, height)

    companion object {
        val EMPTY = Bounds(0, 0, 0, 0)
    }
}

/**
 * One run of text on screen, with where it sits.
 *
 * Produced identically by every [com.rideguard.domain.parse.ScreenSource]:
 *  - the AccessibilityService fills [viewId] and leaves [confidence] at 1.0
 *  - the OCR source leaves [viewId] null and sets a real [confidence]
 *
 * Everything downstream is blind to which one produced it. That is the whole
 * point of the abstraction.
 */
@Serializable
data class TextBlock(
    val text: String,
    val bounds: Bounds = Bounds.EMPTY,
    val viewId: String? = null,
    val confidence: Float = 1f,
) {
    /**
     * Rough proxy for font size. The accessibility tree does not expose text
     * size, but a taller node almost always means bigger type — which is how
     * we tell the headline fare from the small print.
     */
    val glyphHeight: Int get() = bounds.height
}

/** One reading of the screen, from any source. */
@Serializable
data class ScreenSnapshot(
    val packageName: String,
    val blocks: List<TextBlock>,
    val capturedAtMs: Long,
) {
    /** All text joined in reading order — handy for logging and regex fallbacks. */
    val flatText: String by lazy {
        blocks.joinToString(separator = "\n") { it.text }
    }

    /** Blocks sorted top-to-bottom, then left-to-right. Reading order. */
    val inReadingOrder: List<TextBlock> by lazy {
        blocks.sortedWith(compareBy({ it.bounds.top }, { it.bounds.left }))
    }

    fun isEmpty(): Boolean = blocks.none { it.text.isNotBlank() }
}

File: ios/RideGuard/Core/Model/Screen.swift
import Foundation

/// Plain rectangle in screen pixels. Deliberately NOT `CGRect` —
/// this module must stay UI-framework-free so it can be tested without a
/// simulator, exactly like the Kotlin `:domain` module it mirrors.
///
/// Integer pixels rather than points, because the only producer on iOS is
/// Vision reading a screenshot, and screenshots are measured in pixels.
public struct Bounds: Equatable, Hashable, Sendable, Codable {
    public let left: Int
    public let top: Int
    public let right: Int
    public let bottom: Int

    public init(left: Int, top: Int, right: Int, bottom: Int) {
        self.left = left
        self.top = top
        self.right = right
        self.bottom = bottom
    }

    public var width: Int { right - left }
    public var height: Int { bottom - top }
    public var centerX: Int { (left + right) / 2 }
    public var centerY: Int { (top + bottom) / 2 }
    public var area: Int { max(0, width) * max(0, height) }

    public static let empty = Bounds(left: 0, top: 0, right: 0, bottom: 0)
}

/// One run of text on screen, with where it sits.
///
/// On Android this is produced either by the AccessibilityService (which fills
/// `viewId` and leaves `confidence` at 1.0) or by OCR (no `viewId`, a real
/// `confidence`). On iOS only the OCR shape is reachable — see
/// `docs/ios-platform-limits.md` — but the field stays so the two codebases
/// keep the same type and recorded fixtures stay interchangeable between them.
public struct TextBlock: Equatable, Hashable, Sendable, Codable {
    public let text: String
    public let bounds: Bounds
    public let viewId: String?
    public let confidence: Float

    public init(
        text: String,
        bounds: Bounds = .empty,
        viewId: String? = nil,
        confidence: Float = 1
    ) {
        self.text = text
        self.bounds = bounds
        self.viewId = viewId
        self.confidence = confidence
    }

    /// Rough proxy for font size. Neither the accessibility tree nor Vision
    /// reports text size, but a taller run almost always means bigger type —
    /// which is how we tell the headline fare from the small print.
    public var glyphHeight: Int { bounds.height }
}

/// One reading of the screen, from any source.
///
/// `flatText` and `inReadingOrder` are computed once in `init` and stored.
/// Kotlin gets this from `by lazy`; Swift structs cannot lazily memoise
/// without becoming mutating, and every parser touches both properties more
/// than once, so eager computation keeps the cost identical to Android's.
public struct ScreenSnapshot: Equatable, Sendable {
    public let packageName: String
    public let blocks: [TextBlock]
    public let capturedAtMs: Int64

    /// All text joined in reading order — handy for logging and regex fallbacks.
    public let flatText: String

    /// Blocks sorted top-to-bottom, then left-to-right. Reading order.
    public let inReadingOrder: [TextBlock]

    public init(packageName: String, blocks: [TextBlock], capturedAtMs: Int64 = 0) {
        self.packageName = packageName
        self.blocks = blocks
        self.capturedAtMs = capturedAtMs
        self.flatText = blocks.map(\.text).joined(separator: "\n")

        // Kotlin's `sortedWith` is a stable sort and the leg-assignment
        // heuristic depends on it: two blocks on the same baseline must keep
        // the order the source emitted them in. Swift's `sorted` gives no such
        // guarantee, so we break ties on the original index by hand.
        self.inReadingOrder = blocks.enumerated()
            .sorted { a, b in
                if a.element.bounds.top != b.element.bounds.top {
                    return a.element.bounds.top < b.element.bounds.top
                }
                if a.element.bounds.left != b.element.bounds.left {
                    return a.element.bounds.left < b.element.bounds.left
                }
                return a.offset < b.offset
            }
            .map(\.element)
    }

    public var isEmpty: Bool {
        !blocks.contains { !$0.text.allSatisfy(\.isWhitespace) && !$0.text.isEmpty }
    }
}
